Persisting/seeding a routing table

aarshkshah1992 opened the pull request:
Based on PR https://github.com/libp2p/go-libp2p-kad-dht/pull/315 by @raulk 

Please refer to issues: 
#254 
#295

aarshkshah1992 commented:
@raulk  Please take a look:

1) Have refactored the randomSeeder
2) Added tests for the randomSeeder & the snapshot impl
3) Init DHT with the above & use  dht.DefaultBootStrap peers as fallback peers if the corresponding options aren't set

**TODO**
1) An end-end test that proves that the dht gets seeded from a snapshot from the previous run
2) Resolve the import conflicts

aarshkshah1992 commented:
@raulk Reminder to take a look.

aarshkshah1992 commented:
@raulk  Reminder to take a look :) Thanks !

aarshkshah1992 commented:
@raulk  #384  now depends on this. We will populate the RT using the seeder with Default/known bootstrap peers whenever RT becomes empty.

raulk commented:
Hey @aarshkshah1992, I've spoken to @bigs and he's gonna pick this up, as I'm currently tackling a deadline and I don't want to leave this out to dry. Thanks for your amazing work so far. Expect to hear from @bigs shortly.

aarshkshah1992 commented:
@bigs  Hey, this PR is ripe for review. Please take a look when you can. Thanks :)

bigs commented:
will do in a few 👌🏻

On Wed, Sep 4, 2019 at 22:07 Aarsh Shah <notifications@github.com> wrote:

> @bigs <https://github.com/bigs> Hey, this PR is ripe for review. Please
> take a look when you can. Thanks :)
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/libp2p/go-libp2p-kad-dht/pull/383?email_source=notifications&email_token=AABUCWQWTE7ULPL5APC7RALQIBSX5A5CNFSM4IK5YVWKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD55SPMY#issuecomment-528164787>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABUCWTA4GHSRXAPMSNL6A3QIBSX5ANCNFSM4IK5YVWA>
> .
>

aarshkshah1992 commented:
Hey @bigs, reminder to review :)

raulk commented:
@bigs how can I help move this forward?

bigs commented on dht_bootstrap.go:
```diff
@@ -7,14 +7,16 @@ import (
 	u "github.com/ipfs/go-ipfs-util"
-	peer "github.com/libp2p/go-libp2p-peer"
+	peer2 "github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-peer"
```
remove this and just use go-libp2p-core/peer

bigs commented on dht_test.go:
```diff
@@ -22,16 +24,16 @@ import (
 	pb "github.com/libp2p/go-libp2p-kad-dht/pb"
 
-	cid "github.com/ipfs/go-cid"
+	"github.com/ipfs/go-cid"
```
let's restore these aliases. makes it easier for people to track down

raulk replied:
We generally order imports this way:

```go
import (
    [stdlib...]

    [libp2p...] (without aliases)

    [multiformats...]

    [everything else in groups, e.g. all ipfs imports together...]
)
```

bigs commented on persist/seeder.go:
```diff
@@ -0,0 +1,160 @@
+	})
+
+	left := rs.target
+	addPeer := func(p peer.ID) { // adds a peer to the routing table and decrements the left counter if successful.
```
~let's just inline this to the loop~

disregard

bigs commented on persist/seeder.go:
```diff
@@ -0,0 +1,160 @@
+		defer cancel()
+
+		// start dialing
+		sempahore := make(chan struct{}, NSimultaneousDial)
```
```suggestion
		semaphore := make(chan struct{}, NSimultaneousDial)
```

this is really clever :)

bigs commented on persist/seeder.go:
```diff
@@ -0,0 +1,160 @@
+					addPeer(res.p)
+				}
+
+			case <-todo.Done():
```
will this trigger on success?

aarshkshah1992 replied:
@bigs 

```
for left > 0 {
			select {
			case res := <-resCh:
				todo.Decrement(1)
                                   addPeer(res.p)
```

The `addPeer` here decrements the left counter by 1. So, if a result on the channel results in a left value of 0 after decrementing the todo Counter by 1, the loop will exit before `todo.Done()` fires. So no, this will not trigger on a success.

aarshkshah1992 commented on dht_test.go:
```diff
@@ -361,6 +365,71 @@ func TestSearchValue(t *testing.T) {
 	}
 }
 
+func TestSnapshottingAndSeeding(t *testing.T) {
```
@bigs  We have done that in persist/seeder_test.go & persist/snapshot_test.go. Let me know if you have something else in mind.

aarshkshah1992 commented:
@bigs  I have fixed the conflicts & addressed your comments. Thanks for the review ! :) Let me know if we need more changes before merging this.

aarshkshah1992 commented:
Hey @bigs, please take a look when you can. This is very close to getting merged.

raulk commented:
Thanks for all the hard work here! Let me take a look before we merge.

aarshkshah1992 commented:
Thanks a lot @bigs ! Really appreciate it :)

aarshkshah1992 commented:
Hey @raulk 

Just fixed another merge conflict. Please take a look at this when you can. I don't know if the world deserves a snapshotted & seeded Dht, but it surely needs one :)

raulk commented on dht_bootstrap.go:
```diff
@@ -37,6 +38,12 @@ func init() {
+		if err != nil {
+			panic(err)
+		}
+		DefaultBootstrapPeerIDs = append(DefaultBootstrapPeerIDs, info.ID)
```
`DefaultBootstrapPeerIDs` should be private and derived from the `DefaultBootstrapPeers`, once the latter is completely initialized.

raulk commented on go.mod:
```diff
@@ -10,19 +10,26 @@ require (
 	github.com/mr-tron/base58 v1.1.2
 	github.com/multiformats/go-multiaddr v0.0.4
-	github.com/multiformats/go-multiaddr-dns v0.0.3
+	github.com/multiformats/go-multiaddr-dns v0.0.2
```
Why these downgrades?

aarshkshah1992 replied:
This is a mistake. I'll fix it.

raulk commented on persist/interfaces.go:
```diff
@@ -0,0 +1,26 @@
+var logSeed = log.Logger("dht/seeder")
+var logSnapshot = log.Logger("dht/snapshot")
+
+type Seeder interface {
```
Missing godoc on the exported type.

raulk commented on persist/seeder.go:
```diff
@@ -0,0 +1,160 @@
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-kbucket"
+	inet "github.com/libp2p/go-libp2p-net"
```
Deprecated type. Should point to `github.com/libp2p/go-libp2p-core/network`.

raulk commented on persist/seeder.go:
```diff
@@ -0,0 +1,160 @@
+var TotalSeedDialGracePeriod = 30 * time.Second
+
+// NSimultaneousDial is the number of peers we will dial simultaneously
+var NSimultaneousDial = 50
```
Can we call this `SeederConcurrentDials`?

raulk commented on persist/seeder.go:
```diff
@@ -0,0 +1,160 @@
+)
+
+// SeedDialGracePeriod is the grace period for one dial attempt
+var SeedDialGracePeriod = 5 * time.Second
```
Can we call this `SeederDialTimeout`?

raulk commented on persist/seeder.go:
```diff
@@ -0,0 +1,160 @@
+		return attemptSeedWithPeers(fallback)
+	}
+
+	// There is a God after all
```
Saucy, but we should probably dial down the irony here ;-)

raulk commented on persist/seeder.go:
```diff
@@ -0,0 +1,160 @@
+		defer cancel()
+
+		// start dialing
+		semaphore := make(chan struct{}, NSimultaneousDial)
```
This kind of throttling logic already exists in the dialer. Any motivation to put it here too?

aarshkshah1992 replied:
@raulk I see. Do you mean to say that the dialer restricts the number of dials to 'X' and queues the remaining ? If yes, please can you explain your comment here :

https://github.com/libp2p/go-libp2p-kad-dht/pull/384/files/00fffba0aa6948e7549752197f63b85f293a66e9#r334237930
 
What do you mean by 

```
We should stagger these lookups, instead of launching them all at once. 
```

raulk replied:
* Dialer reference: Sure, have a look at: https://github.com/libp2p/go-libp2p-swarm/blob/master/limiter.go.
* Staggering clarification: I responded in #384 directly.

aarshkshah1992 replied:
I realize that this & 384 are different ideas. Sorry for mixing them up.

I had a look at the dial limiter & understand that we queue all TCP dial requests once the file descriptor limit is hit. However, my concern is that we will eat up a huge number of fd's for ourselves if we have too many candidates in the snapshot & do not have some form of rate limiting here. This in turn will slow down dial requests from other parts of the application. Let me know what you think.

raulk commented on persist/seeder.go:
```diff
@@ -0,0 +1,160 @@
+	}
+
+	// attempts to dial to a given peer to verify it's available
+	dialFn := func(ctx context.Context, p peer.ID, res chan<- result) {
```
I'm having second thoughts about putting this dialing logic here. It's probable that all seeders would want to verify they can _actually_ add a peer before doing so. The dialing logic needs to be somewhere common.

I wonder if the seeder should be dumb, poll-driven and not perform any dials. We can move the dialing logic to the DHT, and pass in the current state of the routing table for the seeder to use when making decisions about which candidates to return next.

```go
type Seeder interface {
   // Next returns a slice of peers to attempt to add next, based on the current
   // state of the routing table. Returning a nil slice signals that the seeder
   // is done proposing peers.
   Next(rt *kbucket.RoutingTable) []peer.ID
}
```

aarshkshah1992 replied:
@raulk  

I don't see the value in making the seeder poll driven & moving the burden of actually seeding the RT on the caller. Can you explain why that'd be a good thing ? Also, why should we remove the candidate/fallback peers from the interface ? The set of fallback peers is already configurable which means the caller can pass in peers obtained from any source.

However, I do agree that the dialing logic is pretty generic & can be pulled out. We can put stuff like that in a helpers package that makes it easy for users to construct their own seeder implementations.

raulk replied:
In a nutshell, I'm worried about putting dialling functionality in the Seeder. It simply doesn't belong there and it complicates the seeder, which is just supposed to seed, not dial. I see this being a future footgun, as it's essentially leaking an abstraction and turning into "spooky action at a distance". I did have it like this in my WIP work, so I totally understand that you followed suit, but that was mostly a placeholder -- I was supposed to revisit that aspect at some point.

Solutions I can think of:
1. The Seeder acts like a "proposer", and we call it iteratively until we exhaust the candidates, and the Seeder returns no more proposals. Imagine we have peers [A-E] available.
    * The Seeder implements `Next(rt *RoutingTable, candidates []peer.ID, fallback []peer.ID) (proposed []peer.ID)`.
    * On a first iteration, we call `Next()` with an empty routing table, all candidates ([A-E]), and the fallback peers.
    * The Seeder returns A and B.
    * The DHT bootstrap logic verifies that A works, B is undiallable. It adds A to the routing table, and drops B from the candidate set.
    * We call `Next()` with the routing table with A, and candidates [C-E] (having dropped B).
    * This goes on until the candidate set is exhausted, and the dialer returns nil.
2. Pass in a "validate function" to the seeder, owned by the DHT. This func would perform the dial. The Seeder calls this function and can therefore verify if the peer is diallable without owning the logic.

I still prefer 1. WDYT?

aarshkshah1992 replied:
@raulk  I implemented **1** to get a feel for what you were saying & it is a good idea indeed. Now that we have decoupled 'proposing' seeders from 'seeding', users can simply plugin custom proposing strategies (preferring candidates in ascending order of XOR distance from self for example) & the seeding will out of the box. Thanks for this !

aarshkshah1992 replied:
@raulk I implemented 1 to get a feel for what you were saying & it is a good idea indeed. Now that we have decoupled 'proposing' seeders from 'dialling' & 'seeding' the RT, users can simply plug in custom 'proposal' strategies (one strategy could be preferring candidates in ascending order of XOR distance from self for example) & the 'dialling'/'seeding' will just work out of the box. Thanks for this !

aarshkshah1992 commented on dht.go:
```diff
@@ -89,17 +90,53 @@ func New(ctx context.Context, h host.Host, options ...opts.Option) (*IpfsDHT, er
+	}
+
+	// schedule periodic snapshots
+	sproc := periodicproc.Tick(cfg.Persistence.SnapshotInterval, func(proc goprocess.Process) {
```
I see what you mean. So, in it's current form, this code could persist an empty RT if the snapshotting go-routine fires after all peers have been dropped from the RT, but Dht hasn't been closed yet.

However, this can also prove to be a blessing in disguise because storing an empty RT & then seeding with bootstrap peers after we restart could save us from ALWAYS storing a poisoned RT if an attacker messed up our RT and found a way to immediately shut us down.

So, let's see how this works in practise & fix it if required ?

aarshkshah1992 commented:
@raulk  Thank you so much for the review. Just some questions/concerns around some comments. Please can you take a look and I'll get the changes in myself :)

aarshkshah1992 commented:
Hey @raulk ,

Have fixed the conflicts & addressed all your comments except for these for which I'll wait for your reply:

1.[ Changing the seeder interface & moving the dialer out of the seeder](https://github.com/libp2p/go-libp2p-kad-dht/pull/383#discussion_r334237289) 
2. [ Removing the dialer throttling logic] (https://github.com/libp2p/go-libp2p-kad-dht/pull/383#discussion_r334237157)
3. [Storing the routing table before it is cleared](https://github.com/libp2p/go-libp2p-kad-dht/pull/383#discussion_r334236444)

aarshkshah1992 commented on dht.go:
```diff
@@ -225,7 +227,7 @@ func makeDHT(ctx context.Context, h host.Host, cfg *opts.Options, bucketSize int
 			if dht.routingTable.Size() == 0 {
 				logger.Infof("rt recovery proc: received request with reqID=%s, RT is empty. initiating recovery", req.id)
-				// TODO Call Seeder with default bootstrap peers here once #383 is merged
+				// TODO Call SeedsProposer with default bootstrap peers here once #383 is merged
```
need to fix this comment. SeedsProposer should be "dht.seedRoutingTable"

aarshkshah1992 commented:
Hey @raulk ,

Have addressed all your comments except one. Please take a look when you can :)
The only thing I am still not convinced of is the comment [here](https://github.com/libp2p/go-libp2p-kad-dht/pull/383#discussion_r335039717).

meyer9 commented on dht_rt_seeder.go:
```diff
@@ -0,0 +1,119 @@
+				for _, p := range peers {
+					semaphore <- struct{}{}
+					go func(p peer.ID, res chan<- result) {
+						dialFn(ctx, p, resCh)
```
Shouldn't this pass `res`, not resCh? Not sure it really matters in this case, but either that or remove the unused argument.

aarshkshah1992 replied:
Agreed. Thanks for spotting this.

meyer9 commented on dht_rt_seeder.go:
```diff
@@ -0,0 +1,119 @@
+			}(todial)
+
+			// number of results we expect on the result channel
+			todo := todocounter.NewSyncCounter()
```
Why not use sync.WaitGroup or just an int. This isn't threaded afaict.

aarshkshah1992 replied:
Yes, we could use a simple int here to keep a track of remaining Dials & exit the loop when it goes to zero. A `todocounter` really shines when the go-routines we create for tasks(dial attempts for example) create more tasks themselves.

https://github.com/ipfs/go-todocounter/blob/742667602a47ab3a2b7f17d935019c3255719dce/counter.go#L7

But, we do NOT have that here as a dial attempt does not generate one more peer we need to dial to.

aarshkshah1992 commented on dht_rt_seeder.go:
```diff
@@ -0,0 +1,119 @@
+
+// SeederDialTimeout is the total grace period for a group of dial attempts
+var SeederDialTimeout = 30 * time.Second
+
```
Have mixed up the naming of the above two vars. Need to fix.

aarshkshah1992 commented:
Hey @raulk ,

I have [decoupled](https://github.com/libp2p/go-libp2p-kad-dht/pull/383#discussion_r335034432) seed proposal from seeding. Thanks for the cool idea.

All your comments except for one have been addressed. Please take a look when you can :)

The only one I am still not sure of is removing the rate limiting from the seeds dialling logic. I've explained it in this [comment](https://github.com/libp2p/go-libp2p-kad-dht/pull/383#discussion_r335039717). Let me know what you think.

Also, there are some minor nits I've pointed out that I'll fix after this round of review. Thanks !

aarshkshah1992 commented on dht_rt_seeder.go:
```diff
@@ -0,0 +1,119 @@
+					if res.err != nil {
+						logger.Infof("dht rt seeder: discarded proposed peer due to dial error; peer ID: %s, err: %s", res.p, res.err)
+					} else {
+						if _, err := dht.routingTable.Update(res.p); err != nil {
```
Need to document  the fact that till #283 is fixed, this could result into two calls to `rt.Update(P)` as dialling & connecting to `P` will also cause a`rt.Update(p)` call if the dht notification service has already started.

However, this does not cause any unwanted side effects for now.

aarshkshah1992 commented:
Hey @raulk ,

This one is incredibly close to getting merged & will also help make the world a better place, in addition to unlocking #387 (which should get done fairly quickly, we already have the algorithm for that one in place).

Please please take a look if you can !

aschmahmann commented on persist/seeds_proposer_test.go:
```diff
@@ -0,0 +1,129 @@
+		"No Proposals -> RT already has target peers": {
+			10, 10, 0, 0, 10, 0,
+			0,
+		},
```
can condense the spacing for these

aarshkshah1992 commented on persist/snapshot.go:
```diff
@@ -0,0 +1,86 @@
+		return err
+	}
+
+	return dsp.Put(dsSnapshotKey, bytes)
```
Good point. I think a synchronous write is better here as it maps nicely to what we are trying to do here i.e. recovery after a dht crash.

aarshkshah1992 replied:
Done.

aarshkshah1992 commented on dht.go:
```diff
@@ -89,18 +92,58 @@ func New(ctx context.Context, h host.Host, options ...opts.Option) (*IpfsDHT, er
+		// should never happen
+		if err != nil {
+			logger.Error("failed to initialize the default datastore backed snapshotter")
+			panic(err)
```
Nope, let's return an error here.

aarshkshah1992 commented on dht.go:
```diff
@@ -89,18 +92,58 @@ func New(ctx context.Context, h host.Host, options ...opts.Option) (*IpfsDHT, er
+		cfg.Persistence.FallbackPeers = getDefaultBootstrapPeerIDs()
+	}
+
+	dht := makeDHT(ctx, h, &cfg, cfg.BucketSize)
```
I agree, this is definitely cleaner.

aarshkshah1992 commented on dht_rt_seeder_test.go:
```diff
@@ -0,0 +1,104 @@
+
+	for name, testcase := range testCases {
+		// create host for self
+		self := bhost.New(swarmt.GenSwarm(t, context.Background(), swarmt.OptDisableReuseport))
```
Agreed

aarshkshah1992 commented on dht_rt_seeder_test.go:
```diff
@@ -0,0 +1,104 @@
+			9},
+
+		"Empty Candidates": {0, 0, 5,
+			5},
```
Done

aarshkshah1992 commented on dht_test.go:
```diff
@@ -12,28 +12,31 @@ import (
 
-	"github.com/ipfs/go-cid"
-	u "github.com/ipfs/go-ipfs-util"
+	"github.com/libp2p/go-libp2p-core/peer"
```
Done

aarshkshah1992 commented on dht_test.go:
```diff
@@ -388,6 +392,67 @@ func TestSearchValue(t *testing.T) {
+	assert.NoError(t, selfDht.Close())
+	peerIDs, err := snapShotter.Load()
+	assert.NoError(t, err)
+	assert.Len(t, peerIDs, 4)
```
Good catch. This race can happen as pointed by @raulk [here](https://github.com/libp2p/go-libp2p-kad-dht/pull/383#discussion_r334236444).

Have fixed this.

aarshkshah1992 commented on persist/interfaces.go:
```diff
@@ -0,0 +1,32 @@
+	// Returns nil if it has no more proposals.
+	// Note: Seeding a routing table with the eligible peers will work only if the the dht uses a persistent peerstore across restarts.
+	// This is because we can recover metadata for the candidate peers only if the peerstore was/is persistent.
+	Propose(rt *kbucket.RoutingTable, candidates []peer.ID, fallback []peer.ID) []peer.ID
```
@aschmahmann

You can take a look at the discussion [here](https://github.com/libp2p/go-libp2p-kad-dht/pull/315#issue-268346676) & [here](https://github.com/libp2p/go-libp2p-kad-dht/pull/383#discussion_r334531196) that made us arrive at this interface. The basic rationale is:

**1.** The proposer is meant to _propose_ that the RT be moved from state X -> state Y by applying some criteria on the set of peers P provided to it (candidates + fallbacks). It needs to look at the current RT state to determine what state it should go to next. 

For eg: I would want to write a proposer that wants to ensure that the RT have peers belonging to as many different buckets(CPL's) so that the RT is aware of as may "regions" in the kad tree as is possible from the set of peers provided to it.  

Basically, because the proposer exists to propose peers to seed the RT, it does make sense for it know the current state of the RT.

**2.** Hmm, we could add the host to the interface but... I think your idea of doing the host/peer store filtering in the caller(dht_rt_seeder.go) itself is better as I don't see why the caller would want to pass in the candidates/fallbacks that do not exist in the peerstore or why a proposer would want to propose peers that do not exist in the peerstore. 

Wdyt ?

aschmahmann replied:
> I would want to write a proposer that wants to ensure that the RT has peers belonging to as many different buckets(CPL's) so that the RT is aware of as may "regions" in the kad tree as is possible from the set of peers provided to it. 

Maybe this was always the intention, but that wasn't quite obvious from @raulk's examples of peer selection from https://github.com/libp2p/go-libp2p-kad-dht/pull/315#issue-268346676

> we can pick candidates randomly, or we can choose based on latency, geographical location (e.g. to favour proximity), etc.

@raulk are we all on the same page that the `Proposer` may utilize the state of the routing table as part of the metric in determining which peers to propose and the ordering of those peers?

I've added some thoughts below https://github.com/libp2p/go-libp2p-kad-dht/pull/383#discussion_r355873456 taking this possibility into account.

aarshkshah1992 commented on persist/seeds_proposer.go:
```diff
@@ -0,0 +1,84 @@
+			if addrs := rs.host.Peerstore().Addrs(p); len(addrs) == 0 {
+				logSeedProposer.Infof("discarding candidate as we no longer have addresses: %s", p)
+				continue
+			}
```
aschmahmann

I have address this comment above. While it does make sense for the RT to be a part of the interface & let the implementation decide how to use it for filtering, the host/peerstore filtering can be done in the caller itself.

aarshkshah1992 commented on persist/seeds_proposer.go:
```diff
@@ -0,0 +1,84 @@
+	if rt.Size() >= rs.target {
+		logSeedProposer.Infof("not returning any proposals as RT already has %d peers and target was %d", rt.Size(), rs.target)
+		return nil
+	}
```
@aschmahmann Good point. The proposer should basically be dumb & keep proposing as long as the set of peers passed to it is non-empty. We let the caller decide if it's target has been achieved.

The target is already configurable/Option with a reasonable default value.

aarshkshah1992 replied:
I agree. 

We let the caller decide when it's _target_ has been achieved & finish up as the notion of a _target_ applies to the caller who is doing the work & not proposing. Will move this to the caller.

Any concerns @raulk ?

aschmahmann commented on dht_rt_seeder.go:
```diff
@@ -0,0 +1,119 @@
+
+	for {
+
+		todial := dht.seedsProposer.Propose(dht.routingTable, peersFromSet(remainingCandidates), peersFromSet(remainingFallbacks))
```
Taking into account https://github.com/libp2p/go-libp2p-kad-dht/pull/383#discussion_r357988342.

I'd like to understand a little better what we're trying to do here. The logic here seems to work fine, but depending on what we're trying to accomplish it might be overly complicated or a little confusing. I'm not recommending any blocking code changes here, but just trying to understand/clarify some of the interface considerations.

1) We propose peers based on how good the peers are (no routing table info)
   - How good peers are probably won't change much during a seeding, so no repeated computation is really required here
   - Why not just return a channel of peers and keep pulling off peers until we've hit the number of seeders we care about?
2) We propose peers based on how good the peers are **AND** the state of the routing table
   - We potentially have two operations that we need to run repeatedly: Propose and Dial
   - If Propose is more expensive then Dial then the current approach works well. We Propose a bunch of peers and Dial some subset (`seederConcurrentDials`) of them at once. Once we've Dialed all the proposed peers we call Propose again and restart this procedure.
        - However, we're not necessarily dialing peers in the optimal order. For example, we really want to fill bucket X with one peer, but that peer dial fails, it would be optimal to try the next peer in that bucket as the next dial but we won't because of the batching
   - If Dial is much more expensive then Propose (I expect this is the case), then it'd be nice if we could initially propose `seederConcurrentDials` peers and then Propose a single new peer whenever a goroutine becomes free. This way we're always dialing the most valuable next peer.
      - We could have something close to this currently by just suggesting state-based implementations of Propose only return a one (or a few) peers at a time. A little more expensive for the first set of dials, but probably not a big deal.
      - If Dial is not much more expensive then Propose then this is wasteful

@aarshkshah1992 am I describing the situation accurately? Does my recommendation for state-based Proposers to return very small sets of peers make sense, or am I misunderstanding how this is supposed to work?

aarshkshah1992 replied:
@aschmahmann  I think you have the nail on the head. Dial can be assumed to be more expensive than propose.

For now, I have made the _Proposer_ return a channel which it will close when it's done proposing. This greatly simplifies the code. 

Please can you elaborate a bit on the advantage of having the proposer returning small peer sets rather than the the whole ordering on the channel like we do in this new implementation ?

aschmahmann replied:
> Is it because the repeated interplay between dial & propose makes it easy for the proposer to observer RT state changes across dials ?

Yes. This is why I asked if we cared about RT state, or just peer quality. If we only care about peer quality (or just use the random proposer) then using a channel is fine. However, the proposer cannot submit the "optimal peer" unless it has the latest routing table information.

While we may still be able to use channels, if we really are just submitting proposed peers one (or a few) at a time we may as well do that with a function call.

aschmahmann commented on dht_rt_seeder.go:
```diff
@@ -0,0 +1,111 @@
+			}
+
+			// attempts to dial to a given peer to verify it's available
+			dialFn := func(ctx context.Context, p peer.ID, res chan<- result) {
```
should be able to define this outside of the for loop

aarshkshah1992 replied:
Problem has been fixed.

aschmahmann commented on dht_rt_seeder.go:
```diff
@@ -0,0 +1,111 @@
+		LOOP:
+			for {
+				select {
+				case res := <-resCh:
```
Why do we wait for all the dials to be done before asking for more proposed peers to dial, doesn't this throttle how many peers we can dial simultaneously?

aarshkshah1992 replied:
This problem has been fixed now.

aschmahmann commented on dht_rt_seeder.go:
```diff
@@ -0,0 +1,111 @@
+	peersFromSet := func(m map[peer.ID]struct{}) []peer.ID {
+		peers := make([]peer.ID, len(m))
+		i := 0
+		for p := range m {
```
nit: you can do `for i, p := range m`

aarshkshah1992 replied:
Not required anymore

aschmahmann commented on dht_rt_seeder_test.go:
```diff
@@ -0,0 +1,100 @@
+		candidateHosts := make([]bhost.BasicHost, testcase.nTotalCandidates)
+		candidatePids := make([]peer.ID, testcase.nTotalCandidates)
+		for i := 0; i < testcase.nTotalCandidates; i++ {
+			h := *bhost.New(swarmt.GenSwarm(t, context.Background(), swarmt.OptDisableReuseport))
```
use ctx

aarshkshah1992 replied:
Done

aschmahmann commented on dht_rt_seeder_test.go:
```diff
@@ -0,0 +1,100 @@
+		fallbackHosts := make([]bhost.BasicHost, testcase.nFallbacks)
+		fallbackPids := make([]peer.ID, testcase.nFallbacks)
+		for i := 0; i < testcase.nFallbacks; i++ {
+			h := *bhost.New(swarmt.GenSwarm(t, context.Background(), swarmt.OptDisableReuseport))
```
might as well use ctx

aarshkshah1992 replied:
Done

aschmahmann commented on dht_rt_seeder_test.go:
```diff
@@ -0,0 +1,100 @@
+
+		// crete dht instance with the mock seedsproposer
+		// make sure the default fallback peers are not in the peerstore so rt remains empty
+		newDht, err := New(context.Background(), self, opts.SeedsProposer(&mockSeedsProposer{}), opts.FallbackPeers([]peer.ID{test.RandPeerIDFatal(t)}))
```
might as well use ctx

aarshkshah1992 replied:
Done

aarshkshah1992 replied:
This can also be resolved.

aschmahmann commented on persist/seeds_proposer_test.go:
```diff
@@ -0,0 +1,114 @@
+
+	for name, testcase := range testCases {
+		// create host for self
+		self := bhost.New(swarmt.GenSwarm(t, context.Background(), swarmt.OptDisableReuseport))
```
Use cancellable context pattern

aarshkshah1992 replied:
Done

aschmahmann commented on persist/seeds_proposer_test.go:
```diff
@@ -0,0 +1,114 @@
+		candidateHosts := make([]bhost.BasicHost, testcase.nTotalCandidates)
+		candidatePids := make([]peer.ID, testcase.nTotalCandidates)
+		for i := 0; i < testcase.nTotalCandidates; i++ {
+			h := *bhost.New(swarmt.GenSwarm(t, context.Background(), swarmt.OptDisableReuseport))
```
use ctx once you have the cancellable context

aarshkshah1992 replied:
Done

aschmahmann commented on persist/seeds_proposer_test.go:
```diff
@@ -0,0 +1,114 @@
+		fallbackHosts := make([]bhost.BasicHost, testcase.nFallbacks)
+		fallbackPids := make([]peer.ID, testcase.nFallbacks)
+		for i := 0; i < testcase.nFallbacks; i++ {
+			h := *bhost.New(swarmt.GenSwarm(t, context.Background(), swarmt.OptDisableReuseport))
```
use ctx once you have the cancellable context

aarshkshah1992 replied:
Done

aarshkshah1992 commented on dht.go:
```diff
@@ -89,18 +92,58 @@ func New(ctx context.Context, h host.Host, options ...opts.Option) (*IpfsDHT, er
+	dht.host.Network().Notify((*netNotifiee)(dht))
+
 	dht.proc.AddChild(dht.providers.Process())
 	dht.Validator = cfg.Validator
```
Done

aarshkshah1992 replied:
Done.

aarshkshah1992 commented:
@aschmahmann Thanks for the really in-depth review. 

**The following changes have been made:**

1. The peerstore & the "ignore if already in RT" filtering has been moved to the caller
2.  Managing the target is now the responsibility of the caller & it is a functional option
3. Batching/polling of the proposer has been replaced with the proposer returning a channel that the caller reads from
4. All the code organization changes

**Open questions now**
1. Why should the proposer return small peers sets ? ([here](https://github.com/libp2p/go-libp2p-kad-dht/pull/383#discussion_r358098681))

@raulk  Please can you take a look at the interface changes & the changes in the random seed proposer & let us know what you think ?

aschmahmann commented on dht.go:
```diff
@@ -72,8 +78,17 @@ type IpfsDHT struct {
+	bootstrapCfg          opts.BootstrapConfig
 	triggerRtRefresh      chan chan<- error
 
+	triggerBootstrap chan struct{}
```
this doesn't seem to be used anywhere

aarshkshah1992 replied:
Oh, looks like the remanent of a rebase. Have removed this.

aschmahmann commented on dht.go:
```diff
@@ -94,31 +109,61 @@ var (
+	// set seedsProposer, snapshotter & fallback peers if not set
+	if cfg.Persistence.SeedsProposer == nil {
+		cfg.Persistence.SeedsProposer = persist.NewRandomSeedsProposer()
+	}
```
We can set this in `Defaults` now

aarshkshah1992 replied:
Done

aschmahmann commented on dht.go:
```diff
@@ -94,31 +109,61 @@ var (
-	dht.host.Network().Notify((*netNotifiee)(dht))
+	if len(cfg.Persistence.FallbackPeers) == 0 {
+		cfg.Persistence.FallbackPeers = getDefaultBootstrapPeerIDs()
+	}
```
we can move this to `Defaults` as well?

aarshkshah1992 replied:
We don't want the opts package to depend on the dht package and the opts package isn't the right place to put the function definition.

aschmahmann replied:
k, I don't feel too strongly about this. Just like keeping all the defaults in one place when we can so they're easier to find 😄.

aarshkshah1992 commented:
Hey @raulk 

Would you be able to take a look at this ?

raulk commented:
This PR has accumulated a ton of history and it's hard to process/review. I'm going to close it and open a new one preserving the commit history, as a clean checkpoint.
